internal/topology/variables/clusterclass_variable_validation.go

```diff
@@ -186,38 +289,254 @@ func validateSchema(schema *apiextensions.JSONSchemaProps, fldPath *field.Path)
+						}
+						if cr.Error != nil {
+							if cr.Error.Type == apiservercel.ErrorTypeRequired {
+								allErrs = append(allErrs, field.Required(fldPath.Child("x-kubernetes-validations").Index(i).Child("rule"), cr.Error.Detail))
+							} else {
+								allErrs = append(allErrs, field.Invalid(fldPath.Child("x-kubernetes-validations").Index(i).Child("rule"), schema.XValidations[i], cr.Error.Detail))
+							}
+						}
+						if cr.MessageExpressionError != nil {
+							allErrs = append(allErrs, field.Invalid(fldPath.Child("x-kubernetes-validations").Index(i).Child("messageExpression"), schema.XValidations[i], cr.MessageExpressionError.Detail))
+						} else if cr.MessageExpression != nil {
+							if cr.MessageExpressionMaxCost > StaticEstimatedCostLimit {
+								costErrorMsg := getCostErrorMessage("estimated messageExpression cost", cr.MessageExpressionMaxCost, StaticEstimatedCostLimit)
+								allErrs = append(allErrs, field.Forbidden(fldPath.Child("x-kubernetes-validations").Index(i).Child("messageExpression"), costErrorMsg))
+							}
+							if celContext.TotalCost != nil {
+								celContext.TotalCost.ObserveExpressionCost(fldPath.Child("x-kubernetes-validations").Index(i).Child("messageExpression"), cr.MessageExpressionMaxCost)
+							}
+						}
+						if cr.UsesOldSelf {
+							if uncorrelatablePath != nil {
+								allErrs = append(allErrs, field.Invalid(fldPath.Child("x-kubernetes-validations").Index(i).Child("rule"), schema.XValidations[i].Rule, fmt.Sprintf("oldSelf cannot be used on the uncorrelatable portion of the schema within %v", uncorrelatablePath)))
+							}
+						}
+					}
+				}
+			}
+		}
+
+		// Return if we found some errors in the CEL validations.
+		if len(allErrs) > 0 {
+			return allErrs
+		}
+
+		// Validate example & enum values via CEL.
+		// OpenAPI validation for example & enum values was already done via: validation.ValidateCustomResource
+		// CEL validation for Default values was already done via: structuraldefaulting.ValidateDefaults
+		celValidator := cel.NewValidator(ss, false, celconfig.PerCallLimit)
+		if celValidator == nil {
+			return append(allErrs, field.Invalid(fldPath.Child("x-kubernetes-validations"), "", "failed to create CEL validator"))
+		}
+		if schema.Example != nil {
+			errs, _ := celValidator.Validate(ctx, fldPath.Child("example"), ss, *schema.Example, nil, celconfig.RuntimeCELCostBudget)
+			allErrs = append(allErrs, errs...)
+		}
+		for i, enum := range schema.Enum {
+			if enum != nil {
+				errs, _ := celValidator.Validate(ctx, fldPath.Child("enum").Index(i), ss, enum, nil, celconfig.RuntimeCELCostBudget)
+				allErrs = append(allErrs, errs...)
+			}
+		}
 	}
 
 	return allErrs
 }
+
+// The following funcs are all duplicated from upstream CRD validation at
+// https://github.com/kubernetes/apiextensions-apiserver/blob/v0.30.0/pkg/apis/apiextensions/validation/validation.go#L1317.
```

Comment:
nit: line number for the link does not seem to be correct :-)

Reply:
```suggestion
// https://github.com/kubernetes/apiextensions-apiserver/blob/v0.30.0/pkg/apis/apiextensions/validation/validation.go#L107.
```

Reply:
Let's get rid of the line number. These are copied from all over this file

Reply:
I'll take this as a follow-up. I have a few follow-up PRs queued up anyway